ArduPlane/navigation.cpp

```diff
@@ -38,13 +40,28 @@ void Plane::loiter_angle_update(void)
     } else if (loiter.sum_cd == 0) {
         // use 1 cd for initial delta
         loiter_delta_cd = 1;
+        loiter.prev_lap_alt_cm = current_loc.alt;
+        loiter.next_sum_lap_cd = 36000;
     } else {
         loiter_delta_cd = target_bearing_cd - loiter.old_target_bearing_cd;
     }
     loiter.old_target_bearing_cd = target_bearing_cd;
     loiter_delta_cd = wrap_180_cd(loiter_delta_cd);
 
     loiter.sum_cd += loiter_delta_cd * loiter.direction;
+    loiter.reached_target_alt |= labs(current_loc.alt - next_WP_loc.alt) < 500;
+
+    // detect loop lap
+    if (labs(loiter.sum_cd) >= loiter.next_sum_lap_cd) {
+
+        // check for scenarios where updrafts can keep you from loitering down indefinitely.
+        // after 3 loops, not much alt changing
+        const int32_t alt_delta_of_last_loiter_lap_cm = labs(current_loc.alt - loiter.prev_lap_alt_cm);
+        loiter.unable_to_acheive_target_alt |= (labs(loiter.sum_cd) > 3*36000) && alt_delta_of_last_loiter_lap_cm < 200;
```

Comment:
Additionally this is poorly behaved where after the first 3 laps and 1 lap can trip this detector as the sum_cd will always be > 3*3600.

The correct fix is to only run this check every 3*3600 cd. It could take slightly longer to detect being stuck, but it doesn't have the above problem which is probably the better solution. We can tolerate being slow to detect this if we are detecting it, but being to sensitive is a bigger problem.